lib/domain.js

```diff
@@ -12,26 +9,43 @@ var getRulesForTld = require('./tld-rules.js');
  * @param {string} host
  * @return {String}
  */
-module.exports = function getDomain (allRules, validHosts, host) {
-  var domain = null, hostTld, rules, rule;
+module.exports = function getDomain(allRules, validHosts, host, isHostClean) {
   var _validHosts = validHosts || [];
+  var cleanHost = cleanHostValue(host, isHostClean);
 
-  if (isValid(_validHosts, host) === false) {
+  if (isValid(_validHosts, cleanHost) === false) {
     return null;
   }
 
-  host = cleanHostValue(host);
-  hostTld = extractTldFromHost(host);
-  rules = getRulesForTld(allRules, hostTld, new Rule({"firstLevel": hostTld, "isHost": _validHosts.indexOf(hostTld) !== -1}));
-  rule = getCandidateRule(host, rules);
+  // Check if `host` ends with '.' followed by one host specified in validHosts.
```

Comment:
I wonder if the trailing dot should be _cleaned_ or kept as part of the clean host value?

If cleaned, I guess the following block can move to the `cleanHost` function.

Reply:
You mean that the `cleanHost` would actually check if the host is part of the `validHosts` and return this value directly right?

Reply:
Hm, I mean the trailing dot thingy appears in various places in the codebase. If the trailing dots are removed as part of the host cleaning process, it simplifies a lot various verification steps. Don't you think?

Reply:
I think the comment is confusing, but it's not the same trailing dot that we are dealing with here. Basically we just check if one element of `validHosts` is a valid general domain of `cleanHost`. But it's not enough to check if `hostname.endsWith(vhost)`, so we need to be sure that either `hostname === vhost` or the character before `hostname.indexOf(vhost)` is `.`. I will add some comment and create a helper function to make it clearer.
```